src/core/basetypes/MCData.cpp

```diff
@@ -510,44 +534,47 @@ Data * Data::dataWithContentsOfFile(String * filename)
         fclose(f);
         return NULL;
     }
-    
-    buf = (char *) malloc((size_t) stat_buf.st_size);
-    
-    read_items = fread(buf, 1, (size_t)  stat_buf.st_size, f);
-    if ((off_t) read_items != stat_buf.st_size) {
-        free(buf);
-        fclose(f);
+
+    unsigned int length = (unsigned int)stat_buf.st_size;
+    void * bytes = mmap(NULL, length, PROT_READ, MAP_PRIVATE, fileno(f), 0);
+    fclose(f);
+
+    if (bytes == MAP_FAILED) {
         return NULL;
     }
-    
-    data = Data::data();
-    data->takeBytesOwnership(buf, (unsigned int) stat_buf.st_size);
-    
-    fclose(f);
-    
-    return data;
+
+    Data * result = new Data();
```

Comment:
Could you go back to:
```
data = new Data::dataWithMmapedData(bytes, length);
```

```
Data * dataWithMmapedData()
{
  data = Data::data();
  data->takeBytesOwnership(buf, (unsigned int) stat_buf.st_size, mmapDeallocator);
}
```

Could you also add a readonly flag to the data when takeBytesOwnership() is used?

And then, we need some fallback implementation, which would make Data back to
malloc'ed blocks when the user of the API tries to call one of the following methods:

```
        virtual void appendData(Data * otherData);
        virtual void appendBytes(const char * bytes, unsigned int length);
        virtual void setBytes(const char * bytes, unsigned int length);
        virtual void setData(Data * otherData);
```